synapse/storage/databases/main/event_federation.py

```diff
@@ -1292,6 +1292,79 @@ def _get_backfill_events(
 
         return event_id_results
 
+    @trace
+    async def record_event_backfill_attempt(self, event_id: str) -> None:
+        if self.database_engine.can_native_upsert:
```

Comment:
aside: we need to remove the `can_native_upsert` flag; all Synapse-supported RDBMSes support it nowadays, or we could reasonably change our compatibility policy to make that true; see https://github.com/matrix-org/synapse/issues/12983.
The only reason we should need to choose to emulate an upsert now would be if the table was 'unsafe to upsert' at this point in time because of an index rebuild.

Reply:
That said: it may be worth adding the condition checking if the table is currently unsafe to upsert. I *doubt* this will bite us ever, but just in case we ever reindex it for some reason.

Reply:
Just to check-in on this point. It sounds like there is nothing for me to do in this PR but a good note for how to clean things up in the future.

Reply:
I think you, for safety/paranoia, want a condition on `"event_failed_backfill_attempts" not in self._unsafe_to_upsert_tables`.
This is just protection in case of future changes — the usual upsert mechanisms use this check automatically, but since you're having to go a bit custom here then we should do the check.

Reply:
> I think you, for safety/paranoia, want a condition on `"event_failed_backfill_attempts" not in self._unsafe_to_upsert_tables`.

Further discussed in the other thread https://github.com/matrix-org/synapse/pull/13589#discussion_r961102072